include/micm/solver/backward_euler.inl

```diff
@@ -0,0 +1,206 @@
+        linear_solver.Factor(state.jacobian_, state.lower_matrix_, state.upper_matrix_, singular);
+
+        auto yn1_iter = Yn1.begin();
+        auto yn_iter = Yn.begin();
+        auto residual_iter = residual.begin();
+        auto forcing_iter = forcing.begin();
+        // forcing_blk in camchem
+        // residual = (Yn1 - Yn) / H - forcing;
+        for(; yn1_iter != Yn1.end(); ++yn1_iter, ++yn_iter, ++residual_iter, ++forcing_iter) {
+          *residual_iter = (*yn1_iter - *yn_iter) / H - *forcing_iter;
+        }
+
+        // the result of the linear solver will be stored in temp
+        // this represnts the change in the solution
+        linear_solver.Solve(residual, temp, state.lower_matrix_, state.upper_matrix_);
+
+        // solution_blk in camchem
+        // Yn1 = Yn1 + temp;
+        // always make sure the solution is positive regardless of which iteration we are on
+        // remove negatives, accept this solution and continue
+        auto temp_iter = temp.begin();
+        yn1_iter = Yn1.begin();
+        for(; temp_iter != temp.end(); ++temp_iter, ++yn1_iter) {
+          *yn1_iter = std::max(0.0, *yn1_iter + *temp_iter);
+        }
+
+        // if this is the first iteration, we don't need to check for convergence
+        if (iterations++ == 0) continue;
+
+        // check for convergence
+        temp_iter = temp.begin();
+        yn1_iter = Yn1.begin();
+
+        // convergence happens when the absolute value of the change to the solution
+        // is less than a tolerance times the absolute value of the solution
+        converged = false;
+        for(; temp_iter != temp.end(); ++temp_iter, ++yn1_iter) {
+          // changes that are much smaller than the tolerance are negligible and we assume can be accepted
+          if (std::abs(*temp_iter) > small) {
+            // std::cout << "temp: " << *temp_iter << " yn1: " << *yn1_iter << " tol: " << tol * std::abs(*yn1_iter) << "\n";
+            if (std::abs(*temp_iter) > tol * std::abs(*yn1_iter)) {
+              converged = false;
+              std::cout << "failed to converge within the newton iteration\n";
+              break;
+            }
+            else { converged = true; }
+          }
+          else { converged = true; }
+        }
+        // correct this check
+      } while(!converged && iterations < max_iter);
+
+      if (!converged) {
+        std::cout << "failed to converge\n";
+        n_successful_integrations = 0;
+
+        if (n_convergence_failures >= time_step_reductions.size()) {
+          // we have failed to converge, accept the solution
+          // TODO: continue on with the current solution to get the full solution
+          n_convergence_failures = 0;
+          // give_up = true;
+          t += H;
+          throw std::system_error(make_error_code(MicmBackwardEulerErrc::FailedToConverge), "Failed to converge");
+          break;
+        };
+
+        Yn = original_Yn;
```

Comment:
We do not need this line since `Yn` is not updated if the solution does not converge.